src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,372 @@
+              break;
+
+            case pal::LabelPosition::QuadrantAbove:
+            case pal::LabelPosition::QuadrantOver:
+            case pal::LabelPosition::QuadrantBelow:
+              labelAlignment = QStringLiteral( "center" );
+              break;
+
+            case pal::LabelPosition::QuadrantAboveRight:
+            case pal::LabelPosition::QuadrantRight:
+            case pal::LabelPosition::QuadrantBelowRight:
+              labelAlignment = QStringLiteral( "left" );
+              break;
+          }
+          break;
+      }
+
+      const double labelRotation = label->getAlpha() * 180 / M_PI;
+
+      const QFont font = labelFeature->definedFont();
+      const QString fontFamily = font.family();
+      const QString fontStyle = font.styleName();
+      const double fontSize = static_cast<double>( font.pixelSize() ) * 72 / context.painter()->device()->logicalDpiX();
+      const bool fontItalic = font.italic();
+      const bool fontBold = font.bold();
+      const double fontLetterSpacing = font.letterSpacing();
+      const double fontWordSpacing = font.wordSpacing();
+
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::MultiLineHeight ) )
+      {
+        QgsTextFormat format = labelSettings.format();
+        format.setLineHeight( dataDefinedValues.value( QgsPalLayerSettings::MultiLineHeight ).toDouble() );
+        labelSettings.setFormat( format );
+      }
+
+      QgsTextFormat format = labelSettings.format();
+      double formatLineHeight = format.lineHeight();
+
+      QgsAttributes attributes;
+      attributes << mMapLayerNames->value( layerId ) << fid << labelText << labelAlignment << labelRotation
+                 << fontFamily << fontStyle << fontSize << fontItalic << fontBold << fontLetterSpacing << fontWordSpacing
+                 << formatLineHeight;
+
+      double x = label->getX();
+      double y = label->getY();
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionX ) )
+      {
+        x = dataDefinedValues.value( QgsPalLayerSettings::PositionX ).toDouble();
+      }
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionY ) )
+      {
+        y = dataDefinedValues.value( QgsPalLayerSettings::PositionY ).toDouble();
+      }
+
+      QgsGeometry geometry( new QgsPoint( x, y ) );
+
+      QgsFeature feature;
+      feature.setAttributes( attributes );
+      feature.setGeometry( geometry );
+      features << feature;
+    };
+
+    QMap<QString, QString> *mMapLayerNames = nullptr;
+    QgsProcessingFeedback *mFeedback = nullptr;
+    QList<QgsFeature> features;
+};
+
+QVariantMap QgsExtractLabelsAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  // MAP_THEME has been taken care of in prepareAlgorithm
+  QgsCoordinateReferenceSystem crs = parameterAsExtentCrs( parameters, QStringLiteral( "EXTENT" ), context );
+  if ( !crs.isValid() )
+    crs = context.project()->crs();
```

Comment:
this should be in prepareAlgorithm (it's potentially not thread safe to access the project crs here, and it's trivial to move to the prepare step)